master/internal/stream/projects.go

```diff
@@ -0,0 +1,264 @@
+) *bun.SelectQuery {
+	q := db.Bun().NewSelect().
+		TableExpr("projects p").
+		Column("p.id").
+		OrderExpr("p.id ASC")
+
+	// add permission scope filter in event of non-global access
+	if !globalAccess {
+		q = permFilterQuery(q, accessScopes)
+	}
+
+	q.WhereGroup(" AND ", func(sq *bun.SelectQuery) *bun.SelectQuery {
+		if len(spec.ProjectIDs) > 0 {
+			q.WhereOr("p.id in (?)", bun.In(spec.ProjectIDs))
+		}
+		if len(spec.WorkspaceIDs) > 0 {
+			q.WhereOr("p.workspace_id in (?)", bun.In(spec.WorkspaceIDs))
+		}
+		return q
+	})
+	return q
+}
+
+// ProjectCollectStartupMsgs collects ProjectMsg's that were missed prior to startup.
+func ProjectCollectStartupMsgs(
+	ctx context.Context,
+	user model.User,
+	known string,
+	spec ProjectSubscriptionSpec,
+) (
+	[]stream.MarshallableMsg, error,
+) {
+	var out []stream.MarshallableMsg
+
+	if len(spec.ProjectIDs) == 0 && len(spec.WorkspaceIDs) == 0 {
+		// empty subscription: everything known should be returned as deleted
+		out = append(out, stream.DeleteMsg{
+			Key:     ProjectsDeleteKey,
+			Deleted: known,
+		})
+		return out, nil
+	}
+	// step 0: get user's permitted access scopes
+	accessMap, err := AuthZProvider.Get().GetProjectStreamableScopes(ctx, user)
+	if err != nil {
+		return nil, err
+	}
+	_, globalAccess := accessMap[model.GlobalAccessScopeID]
+	var accessScopes []model.AccessScopeID
+	// only populate accessScopes if user doesn' have global access
+	if !globalAccess {
+		for id, isPermitted := range accessMap {
+			if isPermitted {
+				accessScopes = append(accessScopes, id)
+			}
+		}
+	}
+
+	// step 1: calculate all ids matching this subscription
+	oldEventsQuery := createFilteredProjectIDQuery(
+		globalAccess,
+		accessScopes,
+		spec,
+	)
+	newEventsQuery := createFilteredProjectIDQuery(
+		globalAccess,
+		accessScopes,
+		spec,
+	)
+
+	// get events that happened prior to since that a relevant (appearance)
+	oldEventsQuery.Where("p.seq <= ?", spec.Since)
+	var exist []int64
+	err = oldEventsQuery.Scan(ctx, &exist)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error when scanning for old offline events: %v\n", err)
+		return nil, err
+	}
+	// and events that happened since the last time this streamer checked
+	newEventsQuery.Where("p.seq > ?", spec.Since)
+	var newEntities []int64
+	err = newEventsQuery.Scan(ctx, &newEntities)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error when scanning for new offline events: %v\n", err)
+		return nil, err
+	}
+
+	// this really should be a union
+	exist = append(exist, newEntities...)
+
+	// step 2: figure out what was missing and what has appeared
+	missing, appeared, err := stream.ProcessKnown(known, exist)
+	if err != nil {
+		return nil, err
+	}
+
+	// step 3: hydrate appeared IDs into full ProjectMsgs
+	var projMsgs []*ProjectMsg
+	if len(appeared) > 0 {
+		query := db.Bun().NewSelect().Model(&projMsgs).Where("project_msg.id in (?)", bun.In(appeared))
+		if !globalAccess {
```

Comment:
Run into `Scan error on column index 5, name "notes": json: cannot unmarshal array into Go value of type stream.JSONB` when returned project has notes

Reply:
Good catch! Corrected, was due to json arrays not be accounted for when I had transitioned the JSONB type over to being more restrictive